imports/server/mediasoup.ts

```diff
@@ -0,0 +1,763 @@
+      createdBy: transport.appData.createdBy,
+    });
+  }
+
+  onProducerCreated(producer: types.Producer) {
+    producer.observer.on('close', Meteor.bindEnvironment(() => {
+      this.producers.delete(producer.appData.producerClient);
+      ProducerServers.remove({ producerId: producer.id });
+    }));
+
+    // Create consumers for other existing transports
+    this.transports.forEach(async (transport, key) => {
+      this.createConsumer(key.split(':')[1], transport, producer);
+    });
+
+    this.producers.set(producer.appData.producerClient, producer);
+    ProducerServers.insert({
+      createdServer: serverId,
+      call: producer.appData.call,
+      peer: producer.appData.peer,
+      transport: producer.appData.transport,
+      producerClient: producer.appData.producerClient,
+      trackId: producer.appData.trackId,
+      producerId: producer.id,
+      createdBy: producer.appData.createdBy,
+    });
+  }
+
+  onConsumerCreated(consumer: types.Consumer) {
+    consumer.observer.on('close', Meteor.bindEnvironment(() => {
+      this.consumers.delete(`${consumer.appData.peer}:${consumer.producerId}`);
+      Consumers.remove({ consumerId: consumer.id });
+    }));
+
+    consumer.observer.on('pause', Meteor.bindEnvironment(() => {
+      Consumers.update({ consumerId: consumer.id }, { $set: { paused: true } });
+    }));
+    consumer.observer.on('resume', Meteor.bindEnvironment(() => {
+      Consumers.update({ consumerId: consumer.id }, { $set: { paused: false } });
+    }));
+
+    this.consumers.set(`${consumer.appData.peer}:${consumer.producerId}`, consumer);
+    Consumers.insert({
+      createdServer: serverId,
+      call: consumer.appData.call,
+      peer: consumer.appData.peer,
+      transportRequest: consumer.appData.transportRequest,
+      transportId: consumer.appData.transportId,
+      producerPeer: consumer.appData.producerPeer,
+      consumerId: consumer.id,
+      producerId: consumer.producerId,
+      kind: consumer.kind,
+      rtpParameters: JSON.stringify(consumer.rtpParameters),
+      paused: consumer.paused,
+      createdBy: consumer.appData.createdBy,
+    });
+  }
+
+  // Mongo callbacks
+
+  peerCreated(peer: PeerType) {
+    Ansible.log('Peer joined call', { peer: peer._id, call: peer.call, createdBy: peer.createdBy });
+    if (!this.callToPeers.has(peer.call)) {
+      this.callToPeers.set(peer.call, new Set());
+    }
+    this.callToPeers.get(peer.call)!.add(peer._id);
+    this.peerToCall.set(peer._id, peer.call);
+
+    // TODO: ideally we only want this to happen on one server to ensure that
+    // oplog events are processed in order, so we need a locking scheme like
+    // "discord-bot" in discord-client-refresher
+    try {
+      Rooms.insert({
+        call: peer.call,
+        routedServer: peer.createdServer,
+        createdBy: peer.createdBy,
+      });
+    } catch (e) {
+      // 11000 is a duplicate key error
```

Comment:
Given how many places we use this, it might be worth extracting a function `isMongoDuplicateKeyError(e)` for use here/below/in locks.ts/puzzle.ts

Reply:
I added a function to `imports/server/mediasoup-api.ts`. I'll leave it as an exercise for future-us to pull it out into a library method and replace the other call sites.